File: python/229求众数 II/229求众数 II.py
"""
给定一个大小为 n 的数组，找出其中所有出现超过 ⌊ n/3 ⌋ 次的元素。

说明: 要求算法的时间复杂度为 O(n)，空间复杂度为 O(1)。

示例 1:

输入: [3,2,3]
输出: [3]
示例 2:

输入: [1,1,1,3,3,2,2,2]
输出: [1,2]

"""

# 大于n/3最多2个数，找到这两个数即可。类似求众数找一个数的方法。
class Solution:
    def majorityElement(self, nums):
        x = 0
        y = 0
        count_x = 0
        count_y = 0
        for num in nums:
            if (count_x == 0 or x == num) and y != num:
                x = num
                count_x += 1
            elif count_y == 0 or y == num:
                y = num
                count_y += 1
            else:
                count_x -= 1
                count_y -= 1
                
        result = []
        c1 = 0
        for num in nums:
            if x == num:
                c1 += 1
        
        if c1 > len(nums) // 3:
            result.append(x)
            
        c2 = 0
        for num in nums:
            if y == num:
                c2 += 1
        
        if c2 > len(nums) // 3 and x != y:
            result.append(y)
        
        return result

File: cpp/229求众数 II/229求众数 II.cpp
/*
给定一个大小为 n 的数组，找出其中所有出现超过 ⌊ n/3 ⌋ 次的元素。

说明: 要求算法的时间复杂度为 O(n)，空间复杂度为 O(1)。

示例 1:

输入: [3,2,3]
输出: [3]
示例 2:

输入: [1,1,1,3,3,2,2,2]
输出: [1,2]

*/

//利用hash表记录每个元素出现次数，再和n/3做比较
#include <vector>
#include <unordered_map>
using namespace std;
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        unordered_map<int, int> m;
        vector<int> result;
        for(int i = 0; i < nums.size(); i++)
        {
            if(m.find(nums[i]) != m.end())
            {
                m[nums[i]]++;
            }
            else
                m[nums[i]] = 1;
        }
        
        int n = nums.size() / 3;
        for(auto it = m.begin(); it != m.end(); it++)
        {
            if(it->second > n)
                result.push_back(it->first);
        }
        
        return result;
    }
};